Astro Databricks provider tutorial (#2245)
astronomer/docs: learn/airflow-databricks.md

sidebar_custom_props: { icon: 'img/integrations/databricks.png' }
---

import CodeBlock from '@theme/CodeBlock';
import databricks_tutorial_dag from '!!raw-loader!../code-samples/dags/airflow-databricks/databricks_tutorial_dag.py';

[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces. Astronomer recommends using Airflow primarily as an orchestrator, and to use an execution framework like Apache Spark to do the heavy lifting of data processing. It follows that using Airflow to orchestrate Databricks jobs is a natural solution for many common use cases.

> Thanks to the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks) you can create Databricks jobs from your Airflow DAG code with full integration and observability, including repair features.

Reviewer: ```suggestion
The easiest way to orchestrate your Databricks notebooks from Airflow and execute them as Databricks Workflows is to use the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks) which provides full observability and control from Airflow so you can manage your workflows from one place.
```

> The easiest way to orchestrate your Databricks notebooks from Airflow and execute them as Databricks Workflows is to use the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks) which provides full observability and control from Airflow so you can manage your Workflows from one place.

Reviewer: ```suggestion
The easiest way to orchestrate your Databricks notebooks from Airflow and execute them as Databricks Workflows is to use the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks), which provides full observability and control from Airflow so you can manage your Workflows from one place.
```

> For orchestration of actions in Databricks outside of Databricks Workflows, the [Databricks provider](https://registry.astronomer.io/providers/databricks) created and maintained by the Airflow community is available. You can find information on the community-managed Databricks provider in the [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow) section.

Reviewer: ```suggestion
For orchestration of actions in Databricks outside of Databricks Workflows, you can use the [Databricks provider](https://registry.astronomer.io/providers/databricks) that is created and maintained by the Airflow community. See more in [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow).
```

> Additionally to the Astro Databricks provider created by Astronomer, a [Databricks provider](https://registry.astronomer.io/providers/databricks) created and maintained by the Airflow community is available. Due to the advanced features available in the Astro Databricks provider, Astronomer recommends its use over the older Databricks provider. For reference you can find information on the community-managed Databricks provider in the [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow) section.

Reviewer: ```suggestion
For orchestration of Databricks tools other than Databricks Workflows, the [Databricks provider](https://registry.astronomer.io/providers/databricks) created and maintained by the Airflow community is available. You can find information on the community-managed Databricks provider in the [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow) section.
```
I _think_ that the astro databricks provider only works for databricks workflows, so this might be a good way to frame this. Unless I'm incorrect, in which case ignore this update.

sidebar_custom_props: { icon: 'img/integrations/databricks.png' }
---

import CodeBlock from '@theme/CodeBlock';
import databricks_tutorial_dag from '!!raw-loader!../code-samples/dags/airflow-databricks/databricks_tutorial_dag.py';

[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces.

The easiest way to orchestrate your Databricks notebooks from Airflow and execute them as Databricks Workflows is to use the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks), which provides full observability and control from Airflow so you can manage your Workflows from one place.

This tutorial covers how to use the Astro Databricks provider for an example use case analyzing renewable energy data. It also discusses why you would want to [use Databricks with Airflow](#why-use-airflow-with-databricks) and [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow) for if the Astro Databricks Provider doesn't fit your use case.

:::info

> All code in this tutorial can be found on [the Astronomer Registry] (INSERT REGISTRY LINK LATER).

Author: ```suggestion
All code in this tutorial can be found on [the Astronomer Registry](https://registry.astronomer.io/dags/renewable_analysis_dag/versions/latest).
```

:::

## Why use Airflow with Databricks

> Many data teams leverage Databricks' optimized Spark engine to run heavy work loads like machine learning models, data transformations, and data analysis. While Databricks offers some orchestration with Databricks Workflows, they are limited in functionality and do not integrate with the rest of your data stack. Using a tool-agnostic orchestrator like Airflow gives you several advantages, like the ability to:

Reviewer 2: ```suggestion
Many data teams leverage Databricks' optimized Spark engine to run heavy workloads like machine learning models, data transformations, and data analysis. While Databricks offers some orchestration with Databricks Workflows, they are limited in functionality and do not integrate with the rest of your data stack. Using a tool-agnostic orchestrator like Airflow gives you several advantages, like the ability to:
```

- Use CI/CD to manage your workflow deployment. Airflow DAGs are Python code, and can be [integrated with a variety of CI/CD tools](https://docs.astronomer.io/astro/ci-cd-templates/template-overview) and [tested](testing-airflow.md).
- Repair single tasks in your Databricks Workflow. If a task fails, you can [re-run it](#repairing-a-databricks-workflow) without having to re-run the entire Workflow, saving valuable compute.
- Use [task groups](task-groups.md) within Databricks Workflows, enabling you to collapse and expand parts of larger Databricks Workflows visually.
- Leverage Airflow [cross-DAG dependencies](cross-dag-dependencies.md) to trigger Databricks Workflows from other DAGs in your Airflow environment, allowing for a data-driven architecture.
- Use familiar Airflow code as your interface to orchestrate Databricks notebooks as Workflows.
- [Inject parameters](#parameters) into your Databricks Workflow at the Workflow-level. These parameters can even be dynamic and retrieved at runtime from other Airflow tasks.

## Time to complete

> This tutorial takes approximately 1 hour to complete.

Author: Haha, those are very rough. Basically 1h for longer ones, 30min for shorter ones so far. Most of the time the tutorial itself takes less but if you need to set up the tools from scratch that can be the bulk of the time like here with Databricks 😅

sidebar_custom_props: { icon: 'img/integrations/databricks.png' }
---

import CodeBlock from '@theme/CodeBlock';
import databricks_tutorial_dag from '!!raw-loader!../code-samples/dags/airflow-databricks/databricks_tutorial_dag.py';

> [Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces. Astronomer recommends using Airflow primarily as an orchestrator, and to use an execution framework like Apache Spark to do the heavy lifting of data processing. It follows that using Airflow to orchestrate Databricks jobs is a natural solution for many common use cases.

Reviewer: ```suggestion
[Databricks](https://databricks.com/) is a popular unified data and analytics platform built around [Apache Spark](https://spark.apache.org/) that provides users with fully managed Apache Spark clusters and interactive workspaces.
```

The easiest way to orchestrate your Databricks notebooks from Airflow and execute them as Databricks Workflows is to use the open source [Astro Databricks provider](https://github.com/astronomer/astro-provider-databricks), which provides full observability and control from Airflow so you can manage your Workflows from one place.

> For orchestration of actions in Databricks outside of Databricks Workflows, you can use the [Databricks provider](https://registry.astronomer.io/providers/databricks) that is created and maintained by the Airflow community. See more in [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow).

Reviewer: ```suggestion
```

> In this tutorial you will learn how to use the Astro Databricks provider with an example use case analyzing renewable energy data.

Reviewer: ```suggestion
This tutorial covers how to use the Astro Databricks provider for an example use case analyzing renewable energy data. It also discusses why you would want to [use Databricks with Airflow]() and [Alternative ways to run Databricks with Airflow](#alternative-ways-to-run-databricks-with-airflow) for if the Astro Databricks Provider doesn't fit your use case.
```
GitHub is being weird and won't let me, but let's add the link to the first section in there.

:::info

> All code in this tutorial can be found on [the Astronomer Registry] (INSERT REGISTRY LINK LATER).

Author: Note as placeholder to put the registry link here once the registry team has done a dev to main :)

:::

> ## Why to use Airflow with Databricks

Reviewer: ```suggestion
## Why use Airflow with Databricks
```

Author: Ahh I was going back and forth on that when writing it haha

> Many data teams like to leverage Databricks' optimized Spark engine to run heavy work loads like machine learning models, data transformations, and data analysis. While Databricks offers some orchestration via Databricks Workflows, it is limited in functionality and does not integrate with the rest of your data stack. Using a tool-agnostic orchestrator like Airflow gives you several advantages, like the ability to:

Reviewer: ```suggestion
Many data teams leverage Databricks' optimized Spark engine to run heavy work loads like machine learning models, data transformations, and data analysis. While Databricks offers some orchestration via Databricks Workflows, it is limited in functionality and does not integrate with the rest of your data stack. Using a tool-agnostic orchestrator like Airflow gives you several advantages, like the ability to:
```

> - have a CI/CD process on your Workflow structures. Airflow DAGs are Python code, and can be [integrated with a variety of CI/CD tools](https://docs.astronomer.io/astro/ci-cd-templates/template-overview) and [tested](testing-airflow.md).

Reviewer: ```suggestion
- Use CI/CD to manage your workflow deployment. Airflow DAGs are Python code, and can be [integrated with a variety of CI/CD tools](https://docs.astronomer.io/astro/ci-cd-templates/template-overview) and [tested](testing-airflow.md).
```

> - repair single tasks in your Databricks Workflow. If a task fails, you can [re-run it](#repairing-a-databricks-workflow) without having to re-run the entire Workflow, saving valuable compute.

Reviewer: ```suggestion
- Repair single tasks in your Databricks Workflow. If a task fails, you can [re-run it](#repairing-a-databricks-workflow) without having to re-run the entire Workflow, saving valuable compute.
```

> - use [task groups](task-groups.md) within Databricks Workflows, enabling you to collapse and expand parts of larger Databricks Workflows visually.

Reviewer: ```suggestion
- Use [task groups](task-groups.md) within Databricks Workflows, enabling you to collapse and expand parts of larger Databricks Workflows visually.
```

> - leverage Airflow [cross-DAG dependencies](cross-dag-dependencies.md) to trigger Databricks Workflows from other DAGs in your Airflow environment, allowing for a data-driven architecture.

Reviewer: ```suggestion
- Leverage Airflow [cross-DAG dependencies](cross-dag-dependencies.md) to trigger Databricks Workflows from other DAGs in your Airflow environment, allowing for a data-driven architecture.
```

> - use familiar Airflow code as your interface to orchestrate Databricks notebooks as Workflows.

Reviewer: ```suggestion
- Use familiar Airflow code as your interface to orchestrate Databricks notebooks as Workflows.
```

> - [inject parameters](#parameters) into your Databricks Workflow at the Workflow-level. These parameters can even be dynamic and retrieved at runtime from other Airflow tasks.

Reviewer: ```suggestion
- [Inject parameters](#parameters) into your Databricks Workflow at the Workflow-level. These parameters can even be dynamic and retrieved at runtime from other Airflow tasks.
```

> All while keeping the benefits of using Databricks Workflows like cheaper Workflow compute, being able to use CI/CD with code of your Notebooks and view Databricks Workflows from the Databricks UI.

Reviewer: ```suggestion
```
I don't think this sentence is really needed here - but if you decide we do need to keep it, I would reword to stand alone so it isn't a "continuation" of the bullets above. e.g. "You can take advantage of these benefits all while...."

Author: I agree it seems a little out of place. Am ok with removing it :)

- Access to a [relational database supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_databases.html). This tutorial uses [PostgreSQL](https://www.postgresql.org/).

## Step 1: Configure your Astro project

An Astro project contains all of the files you need to run Airflow locally.

1. Create a new Astro project:

    ```sh
    $ mkdir astro-databricks-tutorial && cd astro-databricks-tutorial
    $ astro dev init
    ```

2. Add the following packages to your `requirements.txt` file. The [Astro Databricks provider package](https://github.com/astronomer/astro-provider-databricks) and the [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html) are used to load, transform, and analyze, the data. [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/) are used to plot the results.

    ```text
    astro-provider-databricks==0.1.3
    astro-sdk-python==1.6.0
    seaborn==0.12.2
    matplotlib==3.7.1
    ```

> 3. Define the following environment variables in the `.env` file in order to be able to serialize Astro Python SDK and Astro Databricks provider objects:

Reviewer: ```suggestion
3. Define the following environment variable in your `.env` file. This allows you to serialize Astro Python SDK and Astro Databricks provider objects.
```

    ```text
    AIRFLOW__CORE__ALLOWED_DESERIALIZATION_CLASSES = airflow\.* astro\.* astro_databricks\.*
    ```

## Step 2: Prepare the data

> This tutorial uses an Airflow DAG to orchestrate a Databricks job that joins datasets about shares of renewable electricity sources in different years for a selected country and calculates the percentage of electricity coming from solar, wind and hydro power.

Reviewer: ```suggestion
For this tutorial you will analyze two renewable energy datasets to calculate the percentage of electricity coming from solar, wind and hydro power for a given country over time.
```

1. [Download the CSV files](https://github.com/astronomer/learn-tutorials-data/tree/main/databricks-tutorial) from GitHub.
2. Save the downloaded CSV files in the `include` directory of your Astro project.

> This tutorial uses parts of a dataset from a [Kaggle](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) about renewable energy derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).

Reviewer: ```suggestion
This data comes from a [Kaggle dataset](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) about renewable energy derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).
```

## Step 3: Prepare your object storage

Create a new bucket named `databricks-tutorial-bucket` in your object storage solution. In this tutorial we use AWS S3, but you can use any solution that is supported by the Astro SDK.

## Step 4: Create Databricks Notebooks

In this tutorial you will orchestrate a Databricks job that sequentially runs two notebooks. 

1. [Create an empty notebook](https://docs.databricks.com/notebooks/notebooks-manage.html) in your Databricks workspace called `join_data`. 

> 2. Copy and paste the following code into the `join_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials. If you are using a different object storage than AWS S3 you will need to replace the code wrapped in `# --------- AWS S3 specific -------- #` comments with code connecting to your object storage.

Reviewer: ```suggestion
2. Copy and paste the following code into the `join_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials.
```

    # define results table
    schema = StructType(
        [
            StructField("Year", IntegerType(), True),
        ]
    )
    result_df = spark.createDataFrame([], schema)

    # join data tables to result table
    for df in dfs:
        col_name = df.columns[3]

        df = df.select("Year", col_name)
        result_df = result_df.join(df, "Year", "outer")

    # convert spark dataframe to pandas and remove potentially duplicated columns
    pandas_df = result_df.toPandas()
    df_t = pandas_df.T
    df_t = df_t.loc[~df_t.index.duplicated(keep="first")]
    df_clean = df_t.T
    csv_buffer = StringIO()
    df_clean.to_csv(csv_buffer, index=False)

    # --------- AWS S3 specific --------- #
    # upload data as a CSV file to S3
    s3 = boto3.client("s3", aws_access_key_id=ACCESS_KEY, aws_secret_access_key=SECRET_KEY)
    s3.put_object(
        Body=csv_buffer.getvalue(),
        Bucket=BUCKET_NAME,
        Key=f"{S3_FOLDER_JOINED_DATA}/{entity_name}.csv",
    )
    # --------- /AWS S3 specific -------- #
    ```

3. Create a second empty notebook in your Databricks workspace called `transform_data`.

> 4. Copy and paste the following code into the `transform_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials, see the Databricks documentation for recommended ways to securely [manage Secrets](https://docs.databricks.com/security/secrets/index.html) in Databricks. If you are using a different object storage than AWS S3 you will need to replace the code wrapped in `# --------- AWS S3 specific -------- #` comments with code connecting to your object storage.

Reviewer: ```suggestion
4. Copy and paste the following code into the `transform_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials.
```

        df = spark.createDataFrame(csv_reader, schema)
        dfs.append(df)

    # calculate summation column for each Spark dataframe
    dfs_transformed = []
    for df in dfs:
        df = df.withColumn(
            "SHW%",
            col("Solar (% electricity)")
            + col("Hydro (% electricity)")
            + col("Wind (% electricity)"),
        )
        dfs_transformed.append(df)

    # convert each spark dataframe into pandas and load to S3
    for file_name, df in zip(csv_files, dfs_transformed):
        # Convert Spark DataFrame to Pandas DataFrame and write to in memory CSV file
        pandas_df = df.toPandas()
        csv_buffer = StringIO()
        pandas_df.to_csv(csv_buffer, index=False)

        # --------- AWS S3 specific --------- #
        # upload data to S3
        s3 = boto3.client(
            "s3", aws_access_key_id=ACCESS_KEY, aws_secret_access_key=SECRET_KEY
        )
        s3.put_object(
            Body=csv_buffer.getvalue(),
            Bucket=BUCKET_NAME,
            Key=f"{S3_FOLDER_TRANSFORMED_DATA}/{file_name.split('/')[1]}",
        )
        # --------- /AWS S3 specific -------- #

    ```

:::info

> Please note that while you can provide your credentials to AWS in plain text to the notebook code above, this is highly discouraged, especially in a production environment. Databricks offers the possiblility to [configure S3 access with instance profiles](https://docs.databricks.com/aws/iam/instance-profile-tutorial.html) and [secure handling of secrets](https://docs.databricks.com/security/secrets/index.html).

Reviewer: ```suggestion
Providing your credentials to AWS in plain text in your notebook code is highly discouraged in production environments. See the Databricks documentation for recommended ways to securely [manage Secrets](https://docs.databricks.com/security/secrets/index.html) in Databricks for other options.
```

:::

> 

Reviewer: Let's add another `info` box here with the info I removed about updating the notebook code if you're not using S3 for object storage.

## Step 5: Configure connections

This tutorial uses three data tools external to Airflow: Databricks, Amazon S3 (or another object storage) and a PostgreSQL database (or another relational database). You will create an Airflow connection to each of these tools.

1. Start Airflow by running `astro dev start`.

2. In the Airflow UI, go to **Admin** -> **Connections** and click **+**. 

3. Create a new connection named `databricks_conn`. Select the connection type `Databricks` and enter the following information:

    - **Connection ID**: `databricks_conn`.
    - **Connection Type**: `Databricks`.
    - **Host**: Your Databricks host address (format: `https://dbc-1234cb56-d7c8.cloud.databricks.com/`).
    - **Login**: Your Databricks login username (email).
    - **Password**: Your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#databricks-personal-access-tokens).

> 4. Create a new connection named `aws_conn`. Make sure the credentials you add have permission to read and write from your S3 bucket. If you are using a different object storage you will need to adjust this step to connect to your tooling.

Reviewer: ```suggestion
4. Create a new connection named `aws_conn`. Make sure the credentials you add have permission to read and write from your S3 bucket. If you are using a different object storage you will need to adjust this step.
```

    - **Connection ID**: `aws_conn`.
    - **Connection Type**: `Amazon Web Services`.
    - **AWS Access Key ID**: Your [AWS Access Key ID](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html). 
    - **AWS Secret Access Key**: Your AWS Secret Access Key.

5. Create a new connection named `db_conn`. Select the connection type and supplied parameters based on the data warehouse you are using. For a Postgres connection, enter the following information:

    - **Connection ID**: `db_conn`.
    - **Connection Type**: `Postgres`.
    - **Host**: Your Postgres host address.
    - **Schema**: Your Postgres database. 
    - **Login**: Your Postgres login username.
    - **Password**: Your Postgres password.
    - **Port**: Your Postgres port.

:::info

If the right connection type isn't available, you might need to add the [relevant provider package](https://registry.astronomer.io/) to `requirements.txt` and run `astro dev restart`.

:::

> ## Step 6: Write your Airflow DAG

Reviewer: ```suggestion
## Step 6: Create your DAG
```

> The DAG you'll write uses the Astro Databricks provider to create a Databricks Workflow from Airflow tasks referencing the Databricks notebooks you prepared in [Step 4](#step-4-create-databricks-notebooks).

Reviewer: ```suggestion
```

1. In your `dags` folder, create a file called `renewable_analysis_dag.py`.

2. Copy and paste the following DAG code into the file:

    <CodeBlock language="python">{databricks_tutorial_dag}</CodeBlock>

>     This DAG consists of seven tasks and one task group:

Reviewer: ```suggestion
    This DAG uses the Astro Databricks provider to create a Databricks Workflow that will run the two notebooks you prepared in [Step 4](#step-4-create-databricks-notebooks) with the following tasks:
```

>     - The `in_tables` task uses the [`LoadFileOperator`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html) of the Astro Python SDK in order to load CSV files from the local `include` directory to the relational database. This task is [dynamically mapped](dynamic-tasks.md), creating one mapped task instance for each file.

Reviewer: ```suggestion
    - The `in_tables` task uses the Astro Python SDK [`LoadFileOperator`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html) to load CSV files from your local `include` directory to your relational database. This task is [dynamically mapped](dynamic-tasks.md), creating one mapped task instance for each file.
```

>     - The `select_countries` task uses the [`aql.transform`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/transform.html) decorator from the Astro Python SDK to run a SQL query, selecting the relevant rows for `COUNTRY` from each of the temporary tables created by the previous task. The result is stored in another temporary table.

Reviewer: ```suggestion
    - The `select_countries` task uses the Astro Python SDK [`aql.transform`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/transform.html) decorator to run a SQL query selecting the relevant rows for `COUNTRY` from each of the temporary tables created by the previous task. The result is stored in another temporary table.
```

>     - The `save_files_to_S3` task dynamically maps over the country subsets created by the previous task and creates one CSV file per table in your S3 bucket via the [ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html). This task uses the `.map` function, a utility function that can transform XComArg objects, learn more [here](dynamic-tasks.md#transform-outputs-with-map).

Reviewer: ```suggestion
    - The `save_files_to_S3` task uses the [ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html) to dynamically maps over the country subsets created by the previous task and create one CSV file per table in your S3 bucket. This task uses the [`.map` function](dynamic-tasks.md#transform-outputs-with-map), a utility function that can transform XComArg objects.
```

>     - The `databricks_workflow` task group, created using the `DatabricksWorkflowTaskGroup` class, contains two tasks using the `DatabricksNotebookOperator`. This task group automatically creates a Databricks Workflow executing the Databricks notebooks specified in the individual DatabricksNotebookOperators. One of the biggest benefits of this setup is the use of a Databricks job cluster instead of other cluster types, allowing you to [significantly reduce your Databricks cost](https://www.databricks.com/product/pricing). In this tutorial example the tasks within the task group will be the following:

Reviewer: ```suggestion
    - The `databricks_workflow` task group, created using the `DatabricksWorkflowTaskGroup` class, automatically creates a Databricks Workflow executing the Databricks notebooks specified in the individual DatabricksNotebookOperators. One of the biggest benefits of this setup is the use of a Databricks job cluster, allowing you to [significantly reduce your Databricks cost](https://www.databricks.com/product/pricing). The task group contains three tasks:
```

>         - The task group provisions a Databricks `job_cluster` with the spec defined as `job_cluster_spec` and creates the Databricks job from the tasks within the task group. Creation of this cluster and job runs as the automatically generated `launch` task.

Reviewer: ```suggestion
        - The `launch` task, which is automatically generated by the task group, provisions a Databricks `job_cluster` with the spec defined as `job_cluster_spec` and creates the Databricks job from the tasks within the task group.
```

        - The `join_data` task runs the `join_data` notebook in this cluster as the first part of the Databricks job. In this notebook the information from the three input CSVs is joined in one file, using Spark.
        - The `transform_data` task runs the `transform_data` notebook as the second part of the Databricks job. This notebook creates a new column in the data called `"SHW%"` which contains the sum of the percentage of solar, wind and hydro in your country's electricity supply.

    - The `delete_intake_files_S3` task deletes all files from the `country_subset` folder in S3.
    - The `load_file` task retrieves the CSV file that the `transform_data` Databricks notebook wrote to S3 and saves the contents in a temporary table in your relational database.

>     - Tthe `create_graph` task uses the `@aql.dataframe` decorator of the Astro Python SDK to create a graph of the `"SHW%"` column. The graph is saved as a `.png` file in the `include` directory.

Reviewer: ```suggestion
    - The `create_graph` task uses the Astro Python SDK `@aql.dataframe` decorator to create a graph of the `"SHW%"` column. The graph is saved as a `.png` file in your `include` directory.
```

>     - Finally, the `cleanup` task using the [`aql.cleanup`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/cleanup.html) operator of the Astro Python SDK runs in parallel with the whole DAG, cleaning up any temporary tables created by the Astro Python SDK when they are not needed anymore.

Reviewer: ```suggestion
    - Finally, the `cleanup` task uses the Astro Python SDK [`aql.cleanup`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/cleanup.html) operator to clean up any temporary tables created by the Astro Python SDK when they are not needed anymore.
```

3. Run the DAG manually by clicking the play button and view the DAG in the graph view. Double click the task group in order to expand it and see all tasks.  

    ![Astro Databricks DAG graph view](/img/guides/astro_databricks_provider_dag_graph.png)
    
4. Open the `shw.png` file in your `include` folder to see your graph.

    ![SWH graph Switzerland](/img/guides/databricks_tutorial_shw_graph.png)

:::info

If you are using a different object storage you will need to make changes to the DAG code above, namely the `OBJECT_STORAGE_CONN_ID` as well as the operator used in the `delete_intake_files_S3` task. To find relevant operators for your object storage see the [Astronomer registry](https://registry.astronomer.io/).

:::

## How it works

> This section will explain Astro Databricks provider functionality in more depth. You can learn more about the Astro Databricks provider in the [provider documentation](https://astronomer.github.io/astro-provider-databricks/).

Reviewer: ```suggestion
This section explains Astro Databricks provider functionality in more depth. You can learn more about the Astro Databricks provider in the [provider documentation](https://astronomer.github.io/astro-provider-databricks/).
```

> ### Additional parameters

Reviewer: Should this just be "Parameters"? "Additional" makes it sounds like we're showing ones that aren't used at all in the tutorial, but some of them are...

- `job_clusters`: the job clusters to be used in this Workflow. You can provide the full `job_cluster_spec` as shown in the tutorial DAG.
- `notebook_params`: a dictionary of parameters to make available to all notebook tasks in a Workflow.
- `notebook_packages`: a list of dictionaries defining Python packages to be installed in all notebook tasks in a Workflow.
- `extra_job_params`: a dictionary with properties to override the default Databricks Workflow job definitions.

You also have the ability to specifiy parameters at the task level in the DatabricksNotebookOperator:

- `notebook_params`: a dictionary of parameters to make available to the notebook.
- `notebook_packages`: a list of dictionaries defining Python packages to be installed in the notebook.

Note that you cannot specifiy the same packages in the `notebook_packages` parameter of a DatabricksWorkflowTaskGroup and in the `notebook_packages` parameter of a task using the DatabricksNotebookOperator in that same task group. Duplicate entries in this parameter will cause an error within Databricks.

### Repairing a Databricks Workflow

> The Astro Databricks provider includes the possibility to repair a failed Databricks Workflow by making a repair request to the [Databricks Jobs API](https://docs.databricks.com/api-explorer/workspace/jobs/repairrun). Databricks expects a single repair request for all tasks that need to be rerun in one cluster, this can be achieved via the Airflow UI by using the operator extra link **Repair All Failed Tasks**.

Reviewer: ```suggestion
The Astro Databricks provider includes functionality to repair a failed Databricks Workflow by making a repair request to the [Databricks Jobs API](https://docs.databricks.com/api-explorer/workspace/jobs/repairrun). Databricks expects a single repair request for all tasks that need to be rerun in one cluster, this can be achieved via the Airflow UI by using the operator extra link **Repair All Failed Tasks**.
```
Can we rewrite the second sentence here to more clearly explain what the benefit is of using this functionality over rerunning a task the regular way? (I assume it's because using regular rerun would send separate requests and spin up separate clusters?)

![Repair All Failed Tasks OEL](/img/guides/repair_all_failed_databricks_tasks_oel.png)

If you only want to rerun specific tasks within your Workflow you can use the **Repair a single failed task** operator extra link on an individual task in the Databricks Workflow.

![Repair a single failed task OEL](/img/guides/repair_single_failed_databricks_task_oel.png)

## Alternative ways to run Databricks with Airflow

> The Astro Databricks provider is under active development and support for more Databricks task types is still being added. If you want to perform an action in your Databricks environment that is not yet supported by the Astro Databricks provider such as [updating a Databricks repository](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest/modules/DatabricksReposUpdateOperator) you will often find relevant operators in the [community-managed Databricks provider](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest). 

Reviewer: ```suggestion
The Astro Databricks provider is under active development, and support for more Databricks task types is still being added. If you want to orchestrate an action in your Databricks environment that is not yet supported by the Astro Databricks provider such as [updating a Databricks repository](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest/modules/DatabricksReposUpdateOperator), check the [community-managed Databricks provider](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest) for relevant operators. 
```

- Airflow operators. See [Operators 101](what-is-an-operator.md).
- Airflow connections. See [Managing your Connections in Apache Airflow](connections.md).

## Prerequisites

- The [Astro CLI](https://docs.astronomer.io/astro/cli/overview).

> - Access to a Databricks workspace. See [Databricks' documentation](https://docs.databricks.com/getting-started/index.html) for instructions. You can use any workspace that has access to the [Databricks workflows](https://docs.databricks.com/workflows/index.html) feature with a user having permissions to create notebooks and Databricks jobs. You can use any underlying cloud service and a [14-day free trial](https://www.databricks.com/try-databricks) is available.

Reviewer: ```suggestion
- Access to a Databricks workspace. See [Databricks' documentation](https://docs.databricks.com/getting-started/index.html) for instructions. You can use any workspace that has access to the [Databricks Workflows](https://docs.databricks.com/workflows/index.html) feature. You will need a user with permissions to create notebooks and Databricks jobs. You can use any underlying cloud service, and a [14-day free trial](https://www.databricks.com/try-databricks) is available.
```
We should capitalize "Workflows" in Databricks Workflows throughout since that's how they refer to it in their docs

> - Access to an [object storage supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_file.html). This tutorial uses an [AWS S3](https://aws.amazon.com/s3/) bucket, if you are using a different object storage you will need to make changes to the Databricks notebook code and the DAG code that is shown.

Reviewer: ```suggestion
- Access to an [object storage supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_file.html). This tutorial uses an [AWS S3](https://aws.amazon.com/s3/) bucket.
```
Let's leave the note about making changes to the notebook code for the specific section where it's relevant.

- Access to a [relational database supported by the Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/supported_databases.html). This tutorial uses [PostgreSQL](https://www.postgresql.org/).

## Step 1: Configure your Astro project

An Astro project contains all of the files you need to run Airflow locally.

1. Create a new Astro project:

    ```sh
    $ mkdir astro-databricks-tutorial && cd astro-databricks-tutorial
    $ astro dev init
    ```

> 2. Add the [Astro Databricks provider package](https://github.com/astronomer/astro-provider-databricks) and the [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html), as well as the plotting libraries [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/) to your Astro project `requirements.txt` file. This tutorial uses the Astro Python SDK to load and analyze data transformed using a Databricks workflow and seaborn together with matplotlib to produce a graph.

Reviewer: ```suggestion
2. Add the following packaged to your `requirements.txt` file. The [Astro Databricks provider package](https://github.com/astronomer/astro-provider-databricks) and the [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html) are used to load, transform, and analyze, the data. [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/) are used to plot the results.
```

    ```text
    astro-provider-databricks==0.1.3
    astro-sdk-python==1.6.0
    seaborn==0.12.2
    matplotlib==3.7.1
    ```

3. Define the following environment variables in the `.env` file in order to be able to serialize Astro Python SDK and Astro Databricks provider objects:

    ```text
    AIRFLOW__CORE__ALLOWED_DESERIALIZATION_CLASSES = airflow\.* astro\.* astro_databricks\.*
    ```

## Step 2: Prepare the data

This tutorial uses an Airflow DAG to orchestrate a Databricks job that joins datasets about shares of renewable electricity sources in different years for a selected country and calculates the percentage of electricity coming from solar, wind and hydro power.

1. [Download the CSV files](https://github.com/astronomer/learn-tutorials-data/tree/main/databricks-tutorial) from GitHub.
2. Save the downloaded CSV files in the `include` directory of your Astro project.

This tutorial uses parts of a dataset from a [Kaggle](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) about renewable energy derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).

> ## Step 3: Prepare your Object storage

Reviewer: ```suggestion
## Step 3: Prepare your object storage
```

> In you object storage solution create a new bucket. This tutorial uses AWS S3 and a bucket named `databricks-tutorial-bucket`. 

Reviewer: ```suggestion
Create a new bucket named `databricks-tutorial-bucket` in your object storage solution. In this tutorial we use AWS S3, but you can use any solution that is supported by the Astro SDK.
```

## Step 4: Create Databricks Notebooks

In this tutorial you will orchestrate a Databricks job that sequentially runs two notebooks. 

1. [Create an empty notebook](https://docs.databricks.com/notebooks/notebooks-manage.html) in your Databricks workspace called `join_data`. 

> 2. Copy and paste the following code into the `join_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials, see the Databricks documentation for recommended ways to securely [manage Secrets](https://docs.databricks.com/security/secrets/index.html) in Databricks. If you are using a different object storage than AWS S3 you will need to replace the code wrapped in `# --------- AWS S3 specific -------- #` comments with code connecting to your object storage.

Reviewer: ```suggestion
2. Copy and paste the following code into the `join_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials. If you are using a different object storage than AWS S3 you will need to replace the code wrapped in `# --------- AWS S3 specific -------- #` comments with code connecting to your object storage.
```
Let's remove the sentence about databricks secrets from here and step 4 below and add a note at the end of the section saying something like for production you won't want to leave credentials in plain text, see databricks docs for secret management, etc.

    - **Password**: Your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#databricks-personal-access-tokens).

4. Create a new connection named `aws_conn`. Make sure the credentials you add have permission to read and write from your S3 bucket. If you are using a different object storage you will need to adjust this step to connect to your tooling.

    - **Connection ID**: `aws_conn`.
    - **Connection Type**: `Amazon Web Services`.
    - **AWS Access Key ID**: Your [AWS Access Key ID](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html). 
    - **AWS Secret Access Key**: Your AWS Secret Access Key.

5. Create a new connection named `db_conn`. Select the connection type and supplied parameters based on the data warehouse you are using. For a Postgres connection, enter the following information:

    - **Connection ID**: `db_conn`.
    - **Connection Type**: `Postgres`.
    - **Host**: Your Postgres host address.
    - **Schema**: Your Postgres database. 
    - **Login**: Your Postgres login username.
    - **Password**: Your Postgres password.
    - **Port**: Your Postgres port.

:::info

> If a connection type isn't available, you might need to make it available by adding the [relevant provider package](https://registry.astronomer.io/) to `requirements.txt` and running `astro dev restart`.

Reviewer: ```suggestion
If the right connection type isn't available, you might need to add the [relevant provider package](https://registry.astronomer.io/) to `requirements.txt` and run `astro dev restart`.
```

:::

## Step 6: Write your Airflow DAG

The DAG you'll write uses the Astro Databricks provider to create a Databricks workflow from Airflow tasks referencing the Databricks notebooks you prepared in [Step 4](#step-4-create-databricks-notebooks).

> 1. In your `dags` folder, create a file called `renewable_analysis_dag`.

Reviewer: ```suggestion
1. In your `dags` folder, create a file called `renewable_analysis_dag.py`.
```

2. Copy and paste the following DAG code into the file:

    <CodeBlock language="python">{databricks_tutorial_dag}</CodeBlock>

    This DAG consists of seven tasks and one task group:

>     - The `cleanup` task using the [`aql.cleanup`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/cleanup.html) operator of the Astro Python SDK runs in parallel with the whole DAG, cleaning up any temporary tables created by the Astro Python SDK when they are not needed anymore.

Reviewer: I would put this bullet last, since it doesn't matter as much functionality wise :)

    - The `in_tables` task uses the [`LoadFileOperator`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html) of the Astro Python SDK in order to load CSV files from the local `include` directory to the relational database. This task is [dynamically mapped](dynamic-tasks.md), creating one mapped task instance for each file.
    - The `select_countries` task uses the [`aql.transform`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/transform.html) decorator from the Astro Python SDK to run a SQL query, selecting the relevant rows for `COUNTRY` from each of the temporary tables created by the previous task. The result is stored in another temporary table.
    - The `save_files_to_S3` task dynamically maps over the country subsets created by the previous task and creates one CSV file per table in your S3 bucket via the [ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html). This task uses the `.map` function, a utility function that can transform XComArg objects, learn more [here](dynamic-tasks.md#transform-outputs-with-map).

>     - The `databricks_workflow` task group, created using the `DatabricksWorkflowTaskGroup` class, contains two tasks using the `DatabricksNotebookOperator`. 

Reviewer: Let's add a bit more context here on what the astro databricks provider is doing that's "special" - e.g. what is it automating or making easier that you wouldn't be able to do with just the community provider.

    - The `load_file` task retrieves the CSV file that the `transform_data` Databricks notebook wrote to S3 and saves the contents in a temporary table in your relational database.
    - Finally, the `create_graph` task uses the `@aql.dataframe` decorator of the Astro Python SDK to create a graph of the `"SHW%"` column. The graph is saved as a `.png` file in the `include` directory.

3. Run the DAG manually by clicking the play button and view the DAG in the graph view. Double click the task group in order to expand it and see all tasks.  

    ![Astro Databricks DAG graph view](/img/guides/astro_databricks_provider_dag_graph.png)
    
4. Open the `shw.png` file in your `include` folder to see your graph.

    ![SWH graph Switzerland](/img/guides/databricks_tutorial_shw_graph.png)

:::info

If you are using a different object storage you will need to make changes to the DAG code above, namely the `OBJECT_STORAGE_CONN_ID` as well as the operator used in the `delete_intake_files_S3` task. To find relevant operators for your object storage see the [Astronomer registry](https://registry.astronomer.io/).

:::

> ## Alternative ways to run Databricks with Airflow

Reviewer: I think this whole part can be shortened into a couple paragraphs under this header. We definitely don't need to list all the specific hooks and operators since those lists are never up to date for long. Maybe we can focus on use cases where you might still use the community provider, and linking to relevant registry pages/docs/example DAGs.

## Step 1: Configure your Astro project

An Astro project contains all of the files you need to run Airflow locally.

1. Create a new Astro project:

    ```sh
    $ mkdir astro-databricks-tutorial && cd astro-databricks-tutorial
    $ astro dev init
    ```

2. Add the following packages to your `requirements.txt` file. The [Astro Databricks provider package](https://github.com/astronomer/astro-provider-databricks) and the [Astro Python SDK](https://astro-sdk-python.readthedocs.io/en/stable/index.html) are used to load, transform, and analyze, the data. [seaborn](https://seaborn.pydata.org/) and [matplotlib](https://matplotlib.org/) are used to plot the results.

    ```text
    astro-provider-databricks==0.1.3
    astro-sdk-python==1.6.0
    seaborn==0.12.2
    matplotlib==3.7.1
    ```

3. Define the following environment variable in your `.env` file. This allows you to serialize Astro Python SDK and Astro Databricks provider objects.

    ```text
    AIRFLOW__CORE__ALLOWED_DESERIALIZATION_CLASSES = airflow\.* astro\.* astro_databricks\.*
    ```

## Step 2: Prepare the data

> For this tutorial you will analyze two renewable energy datasets to calculate the percentage of electricity coming from solar, wind and hydro power for a given country over time.

Author: I think this works well :)

1. [Download the CSV files](https://github.com/astronomer/learn-tutorials-data/tree/main/databricks-tutorial) from GitHub.
2. Save the downloaded CSV files in the `include` directory of your Astro project.

> This data comes from a [Kaggle dataset](https://www.kaggle.com/datasets/programmerrdai/renewable-energy) about renewable energy derived from [Our World in Data](https://ourworldindata.org/renewable-energy) (License [CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).

Author: Thank you and that is awesome! The idea to try to use these types of examples came up in a climate ERG meeting 😄

## Step 3: Prepare your object storage

Create a new bucket named `databricks-tutorial-bucket` in your object storage solution. In this tutorial we use AWS S3, but you can use any solution that is supported by the Astro SDK.

## Step 4: Create Databricks Notebooks

In this tutorial you will orchestrate a Databricks job that sequentially runs two notebooks. 

1. [Create an empty notebook](https://docs.databricks.com/notebooks/notebooks-manage.html) in your Databricks workspace called `join_data`. 

> 2. Copy and paste the following code into the `join_data` notebook. You can divide the code into cells as you see fit. Make sure to replace the `ACCESS_KEY` and `SECRET_KEY` variables with your respective credentials.

Author: Yes :)

1. Start Airflow by running `astro dev start`.

2. In the Airflow UI, go to **Admin** -> **Connections** and click **+**. 

3. Create a new connection named `databricks_conn`. Select the connection type `Databricks` and enter the following information:

    - **Connection ID**: `databricks_conn`.
    - **Connection Type**: `Databricks`.
    - **Host**: Your Databricks host address (format: `https://dbc-1234cb56-d7c8.cloud.databricks.com/`).
    - **Login**: Your Databricks login username (email).
    - **Password**: Your [Databricks personal access token](https://docs.databricks.com/dev-tools/auth.html#databricks-personal-access-tokens).

4. Create a new connection named `aws_conn`. Make sure the credentials you add have permission to read and write from your S3 bucket. If you are using a different object storage you will need to adjust this step.

    - **Connection ID**: `aws_conn`.
    - **Connection Type**: `Amazon Web Services`.
    - **AWS Access Key ID**: Your [AWS Access Key ID](https://docs.aws.amazon.com/powershell/latest/userguide/pstools-appendix-sign-up.html). 
    - **AWS Secret Access Key**: Your AWS Secret Access Key.

> 5. Create a new connection named `db_conn`. Select the connection type and supplied parameters based on the data warehouse you are using. For a Postgres connection, enter the following information:

Author: You just need a connection to a relational database. Any that is supported by the Astro Python SDK should work and I just picked Postgres as a neutral simple example and tested with it :)

    - The `in_tables` task uses the Astro Python SDK [`LoadFileOperator`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/load_file.html) to load CSV files from your local `include` directory to your relational database. This task is [dynamically mapped](dynamic-tasks.md), creating one mapped task instance for each file.
    - The `select_countries` task uses the Astro Python SDK [`aql.transform`](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/transform.html) decorator to run a SQL query selecting the relevant rows for `COUNTRY` from each of the temporary tables created by the previous task. The result is stored in another temporary table.
    - The `save_files_to_S3` task uses the [ExportToFileOperator](https://astro-sdk-python.readthedocs.io/en/stable/astro/sql/operators/export.html) to dynamically map over the country subsets created by the previous task and create one CSV file per table in your S3 bucket. This task uses the [`.map` function](dynamic-tasks.md#transform-outputs-with-map), a utility function that can transform XComArg objects.

>     - The `databricks_workflow` task group, created using the `DatabricksWorkflowTaskGroup` class, automatically creates a Databricks Workflow executing the Databricks notebooks specified in the individual DatabricksNotebookOperators. One of the biggest benefits of this setup is the use of a Databricks job cluster, allowing you to [significantly reduce your Databricks cost](https://www.databricks.com/product/pricing). The task group contains three tasks:

Reviewer 2: ```suggestion
    - The `databricks_workflow` task group, created using the `DatabricksWorkflowTaskGroup` class, automatically creates a Databricks Workflow that executes the Databricks notebooks you specified in the individual DatabricksNotebookOperators. One of the biggest benefits of this setup is the use of a Databricks job cluster, allowing you to [significantly reduce your Databricks cost](https://www.databricks.com/product/pricing). The task group contains three tasks:
```

>         - The `launch` task, which is automatically generated by the task group, provisions a Databricks `job_cluster` with the spec defined as `job_cluster_spec` and creates the Databricks job from the tasks within the task group.

Reviewer 2: ```suggestion
        - The `launch` task, which the task group automatically generates, provisions a Databricks `job_cluster` with the spec defined as `job_cluster_spec` and creates the Databricks job from the tasks within the task group.
```

        - The `join_data` task runs the `join_data` notebook in this cluster as the first part of the Databricks job. In this notebook the information from the three input CSVs is joined in one file, using Spark.

>         - The `transform_data` task runs the `transform_data` notebook as the second part of the Databricks job. This notebook creates a new column in the data called `"SHW%"` which contains the sum of the percentage of solar, wind and hydro in your country's electricity supply.

Reviewer 2: ```suggestion
        - The `transform_data` task runs the `transform_data` notebook as the second part of the Databricks job. This notebook creates a new column in the data called `"SHW%"`, which contains the sum of the percentage of solar, wind, and hydro in your country's electricity supply.
```

3. Run the DAG manually by clicking the play button and view the DAG in the graph view. Double click the task group in order to expand it and see all tasks.  

    ![Astro Databricks DAG graph view](/img/guides/astro_databricks_provider_dag_graph.png)
    
4. Open the `shw.png` file in your `include` folder to see your graph.

    ![SWH graph Switzerland](/img/guides/databricks_tutorial_shw_graph.png)

:::info

If you are using a different object storage you will need to make changes to the DAG code above, namely the `OBJECT_STORAGE_CONN_ID` as well as the operator used in the `delete_intake_files_S3` task. To find relevant operators for your object storage see the [Astronomer registry](https://registry.astronomer.io/).

:::

## How it works

This section explains Astro Databricks provider functionality in more depth. You can learn more about the Astro Databricks provider in the [provider documentation](https://astronomer.github.io/astro-provider-databricks/).

### Parameters

The DatabricksWorkflowTaskGroup provides configuration options via several parameters:

> - `job_clusters`: the job clusters to be used in this Workflow. You can provide the full `job_cluster_spec` as shown in the tutorial DAG.

Reviewer 2: ```suggestion
- `job_clusters`: the job clusters for this Workflow to use. You can provide the full `job_cluster_spec` as shown in the tutorial DAG.
```

- `notebook_params`: a dictionary of parameters to make available to all notebook tasks in a Workflow.

> - `notebook_packages`: a list of dictionaries defining Python packages to be installed in all notebook tasks in a Workflow.

Reviewer 2: ```suggestion
- `notebook_packages`: a list of dictionaries defining Python packages to install in all notebook tasks in a Workflow.
```

- `extra_job_params`: a dictionary with properties to override the default Databricks Workflow job definitions.

You also have the ability to specifiy parameters at the task level in the DatabricksNotebookOperator:

- `notebook_params`: a dictionary of parameters to make available to the notebook.
- `notebook_packages`: a list of dictionaries defining Python packages to be installed in the notebook.

> Note that you cannot specifiy the same packages in the `notebook_packages` parameter of a DatabricksWorkflowTaskGroup and in the `notebook_packages` parameter of a task using the DatabricksNotebookOperator in that same task group. Duplicate entries in this parameter will cause an error within Databricks.

Reviewer 2: ```suggestion
Note that you cannot specify the same packages in both the `notebook_packages` parameter of a DatabricksWorkflowTaskGroup and the `notebook_packages` parameter of a task using the DatabricksNotebookOperator in that same task group. Duplicate entries in this parameter cause an error in Databricks.
```

### Repairing a Databricks Workflow

The Astro Databricks provider includes functionality to repair a failed Databricks Workflow by making a repair request to the [Databricks Jobs API](https://docs.databricks.com/api-explorer/workspace/jobs/repairrun). Databricks expects a single repair request for all tasks that need to be rerun in one cluster, this can be achieved via the Airflow UI by using the operator extra link **Repair All Failed Tasks**. If you would be using Airflow's built in [retry functionality](rerunning-dags.md#automatically-retry-tasks) a separete cluster would be created for each failed task.

![Repair All Failed Tasks OEL](/img/guides/repair_all_failed_databricks_tasks_oel.png)

> If you only want to rerun specific tasks within your Workflow you can use the **Repair a single failed task** operator extra link on an individual task in the Databricks Workflow.

Reviewer 2: ```suggestion
If you only want to rerun specific tasks within your Workflow, you can use the **Repair a single failed task** operator extra link on an individual task in the Databricks Workflow.
```

![Repair a single failed task OEL](/img/guides/repair_single_failed_databricks_task_oel.png)

## Alternative ways to run Databricks with Airflow

The Astro Databricks provider is under active development, and support for more Databricks task types is still being added. If you want to orchestrate an action in your Databricks environment that is not yet supported by the Astro Databricks provider such as [updating a Databricks repository](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest/modules/DatabricksReposUpdateOperator), check the [community-managed Databricks provider](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest) for relevant operators. 

Additionally, the community-managed Databricks provider contains hooks (for example the [DatabricksHook](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest/modules/DatabricksHook)) that simplify interaction with Databricks, including writing your own [custom Databricks operators](airflow-importing-custom-hooks-operators.md#create-a-custom-operator).

> You can find several example DAGs using the community-managed Databricks provider on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest).

Reviewer 2: ```suggestion
You can find several example DAGs that use the community-managed Databricks provider on the [Astronomer Registry](https://registry.astronomer.io/providers/apache-airflow-providers-databricks/versions/latest).
```

Documentation for both operators can be found on the [Astronomer Registry](https://registry.astronomer.io/providers/databricks?type=operators).

> ### Example - Using Airflow with Databricks

Reviewer: We can remove most of this. If this DAG exists on the registry (or there are any others that use the databricks provider operators), let's just link to that.

For the `DatabricksRunNowOperator`, you only need to provide the `job_id` for the job you want to submit, since the job parameters should already be configured in Databricks. You can find the `job_id` on the **Jobs** tab of your Databricks account. However, you can also provide `notebook_params`, `python_params`, or `spark_submit_params` as needed for your job. In this case, you parameterized your notebook to take in a `Variable` integer parameter and passed in '5' for this example.

> #### Error handling

Reviewer: You can cut all this - it's out of date now :)

For context, when I originally wrote this using Airflow + databricks together was a pretty new concept, so we talked a lot about the mechanics of how that interaction worked. I think it's less novel now, and we can assume this sort of nuance is something users would figure out themselves.

In the case above, you can click on the URL link to get to the Databricks log in order to debug the issue.

> ### Next steps

Reviewer: And we can definitely cut this and the out of date screenshot :)
